design/cassandra/design.md

```diff
@@ -0,0 +1,109 @@
+# Kubernetes Operator for Apache Cassandra
+
+## Why ?
+
+* Cassandra is one of the most popular NoSQL databases.
+* Peer-to-Peer write-anywhere design for very high write and read throughput.
+* Despite those advantages, maintaining a Cassandra cluster is a highly manual and tiring task, which requires extensive familiarity with Cassandra's internals.
+* An operator will automate all the manual interventions that a human administrator normally has to perform and provide a production-ready database out-of-the-box.
+
+## Features
+
+The operator will provide all the important features of Cassandra while exposing a user-friendly declarative UI, in accordance to Kubernetes principles.
+More specifically, those features include:
+
+1. Cluster Creation
+2. (Auto)Scaling 
+3. Failed Cassandra Node Replacement
+4. Schedule Node Repair
+5. Backup/Restore
+6. Metrics endpoint for Kubernetes
+
+## Design
+
+### Overview
+
+* This operator will use the pattern: `CRD` + `Controller` + `Sidecar`
+
+![operator-design-overview](media/operator_overview.jpg)
+
+### Cassandra-Kubernetes Mapping
+
+* The design of the Cassandra CRD will follow the Cassandra terminology as much as possible, in order to be familiar to existing users of Cassandra and not confuse new users studying the Cassandra docs.
+
+* Cassandra abstracts resources in the following order:
+`Cluster` -> `Datacenter` -> `Rack` -> `Node`
+
+* We map those abstractions to Kubernetes in the following way:
+
+|  Cassandra    |  Kubernetes   |
+|   ----        |     ----      |
+|    Cluster    |  Cluster CRD  |
+|   Datacenter  |  StatefulSet  |
```

Comment:
It looks like we are trying to tackle the multi-cluster (federation) problem with this design where we are not just managing the cassandra deployment within just 1 k8s cluster, but across a set of clusters, but please correct me if I'm wrong.  There are some k8s community efforts towards that, but that problem space is still in its infancy.  See https://github.com/kubernetes-sigs/federation-v2 for some of the recent thinking.

Do you see value in managing Cassandra within a single k8s cluster (I personally suspect there is definite value)?  That may be where we need to start for the initial implementation until multi-cluster management makes more headway.

Reply:
I am aware of that fact and right now, the most you can do is run across availability zones. 
However, even managing Cassandra within a single Kubernetes cluster has very high value, as the database is very well suited for write-heavy workloads. 
If we get to the point where we have a self-healing Cassandra Datacenter, then we can look into configuring multiple Cassandra DCs in different K8s clusters without requiring Federation, as the management of each Cassandra DC is mostly independent of what other DCs are doing.

So, to sum up all of the above, the initial end goal is a single self-healing Cassandra DC running in a single Kubernetes Cluster. Multiple Cassandra Racks (usually translated to Availability Zones) should also be supported as a single K8s cluster can span Availability Zones.

Reply:
I think it make sense to break the problem into two parts -- managing Cassandra inside a single Kubernetes cluster, and managing Cassandra across Kubernetes cluster. I think we should start with the former.

Reply:
I agree ! I will be changing the dataCenters key to dataCenter. It also won't be a list anymore. While we plan to support multiple dataCenters in the future, keeping it this way in the beginning will result in a better UX.